flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorPartitionLifecycleTest.java

```diff
@@ -0,0 +1,279 @@
+	}
+
+	@Test
+	public void testPartitionReleaseAfterShutdown() throws Exception {
+		// don't do any explicit release action, so that the partition must be cleaned up on shutdown
+		testPartitionRelease(
+			(jobId, partitionId, taskExecutorGateway) -> { },
+			(jobId, jobAwareShuffleEnvironment) -> { });
+	}
+
+	private void testPartitionRelease(
+		TriConsumer<JobID, ResultPartitionID, TaskExecutorGateway> releaseAction,
+		BiConsumer<JobID, JobAwareShuffleEnvironmentImpl> postReleaseActionAssertion) throws Exception {
+
+		final TaskManagerLocation taskManagerLocation = new LocalTaskManagerLocation();
+
+		final TaskSlotTable taskSlotTable = new TaskSlotTable(
+			Collections.singletonList(ResourceProfile.UNKNOWN),
+			new TimerService<>(TestingUtils.defaultExecutor(), timeout.toMilliseconds()));
+
+		final JobLeaderService jobLeaderService = new JobLeaderService(taskManagerLocation, RetryingRegistrationConfiguration.defaultConfiguration());
+
+		final TaskExecutorLocalStateStoresManager localStateStoresManager = new TaskExecutorLocalStateStoresManager(
+			false,
+			new File[]{tmp.newFolder()},
+			Executors.directExecutor());
+
+		final JobAwareShuffleEnvironmentImpl<?> jobAwareShuffleEnvironment = new JobAwareShuffleEnvironmentImpl<>(new NettyShuffleEnvironmentBuilder().build());
+
+		final TaskManagerServices taskManagerServices = new TaskManagerServicesBuilder()
+			.setTaskManagerLocation(taskManagerLocation)
+			.setTaskSlotTable(taskSlotTable)
+			.setJobLeaderService(jobLeaderService)
+			.setTaskStateManager(localStateStoresManager)
+			.setShuffleEnvironment(jobAwareShuffleEnvironment)
+			.build();
+
+		final CompletableFuture<Void> taskFinishedFuture = new CompletableFuture<>();
+
+		final TestingJobMasterGateway jobMasterGateway = new TestingJobMasterGatewayBuilder()
+			.setRegisterTaskManagerFunction((s, location) -> CompletableFuture.completedFuture(new JMTMRegistrationSuccess(ResourceID.generate())))
+			.setOfferSlotsFunction((resourceID, slotOffers) -> CompletableFuture.completedFuture(slotOffers))
+			.setUpdateTaskExecutionStateFunction(taskExecutionState -> {
+				if (taskExecutionState.getExecutionState() == ExecutionState.FINISHED) {
+					taskFinishedFuture.complete(null);
+				}
+				return CompletableFuture.completedFuture(Acknowledge.get());
+			})
+			.build();
+
+		final TestingTaskExecutor taskManager = createTestingTaskExecutor(taskManagerServices, new HeartbeatServices(Long.MAX_VALUE, Long.MAX_VALUE));
+
+		try {
+			taskManager.start();
+			taskManager.waitUntilStarted();
+
+			final String jobMasterAddress = "jm";
+			rpc.registerGateway(jobMasterAddress, jobMasterGateway);
+
+			// inform the task manager about the job leader
+			jobLeaderService.addJob(jobId, jobMasterAddress);
+			jobManagerLeaderRetriever.notifyListener(jobMasterAddress, UUID.randomUUID());
+
+			while (!taskManagerServices.getJobManagerTable().contains(jobId)) {
+				Thread.sleep(50);
```

Comment:
might be better to give an exit condition to avoid stuck here when other changes cause problem future.